[quant][graphmode][fx] Support quantize per channel in all cases

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43984 [quant][graphmode][fx][api] Call fuse in prepare
* #43983 [quant][graphmode][fx] Support inplace option
* **#44042 [quant][graphmode][fx] Support quantize per channel in all cases**

Summary:
Missed one case last time

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23479345](https://our.internmc.facebook.com/intern/diff/D23479345)

jerryzh168 committed:
[quant][graphmode][fx] Support quantize per channel in all cases

Summary:
Missed one case last time

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support quantize per channel in all cases"


Summary:
Missed one case last time

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support quantize per channel in all cases"


Summary:
Missed one case last time

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23479345](https://our.internmc.facebook.com/intern/diff/D23479345)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support quantize per channel in all cases"


Summary:
Missed one case last time

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23479345](https://our.internmc.facebook.com/intern/diff/D23479345)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support quantize per channel in all cases"


Summary:
Missed one case last time

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23479345](https://our.internmc.facebook.com/intern/diff/D23479345)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support quantize per channel in all cases"


Summary:
Missed one case last time

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23479345](https://our.internmc.facebook.com/intern/diff/D23479345)

[ghstack-poisoned]

vkuzo commented on torch/quantization/fx/utils.py:
```diff
@@ -74,3 +74,58 @@ def graph_pretty_str(g, shorten=True) -> str:
+                return False
+            return True
+
+    def get_next_i(module, qparams):
```
`s/i/qparams_idx/g` for clarity?

vkuzo commented on torch/quantization/fx/utils.py:
```diff
@@ -74,3 +74,58 @@ def graph_pretty_str(g, shorten=True) -> str:
+    return quantize_op, qparams
+
+def quantize_node(parent_module, graph, node, activation_post_process):
+    def noattr(module, qparams, i):
```
`module_has_qparams_attrs_with_idx` or the reverse of that? for clarity

vkuzo commented on torch/quantization/fx/utils.py:
```diff
@@ -74,3 +74,58 @@ def graph_pretty_str(g, shorten=True) -> str:
+        quantize_op = torch.quantize_per_tensor
+    return quantize_op, qparams
+
+def quantize_node(parent_module, graph, node, activation_post_process):
```
any chance we can add a toy example to the docblock?

vkuzo reviewed (approved):
lg, optional naming suggestions inline

supriyar commented on torch/quantization/fx/utils.py:
```diff
@@ -74,3 +74,58 @@ def graph_pretty_str(g, shorten=True) -> str:
+def get_per_tensor_qparams(activation_post_process):
+    assert is_per_tensor(activation_post_process.qscheme), 'Only per tensor quantization is supported'
+    scale, zero_point = activation_post_process.calculate_qparams()
+    scale = float(scale)
+    zero_point = int(zero_point)
```
We also have float qparams supported now so that would be a good idea.

jerryzh168 replied:
the cast is only for per tensor quantization, we should probably change the quantize_per_tensor API to accept tensors instead in the future

supriyar commented on torch/quantization/fx/utils.py:
```diff
@@ -74,3 +74,58 @@ def graph_pretty_str(g, shorten=True) -> str:
+
+def is_per_channel(qscheme):
+    return qscheme == torch.per_channel_affine or \
+        qscheme == torch.per_channel_symmetric
```
Should we also include `per_channel_affine_float_qparams` here?

jerryzh168 replied:
yeah i think so, will add

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support quantize per channel in all cases"


Summary:
Missed one case last time

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23479345](https://our.internmc.facebook.com/intern/diff/D23479345)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support quantize per channel in all cases"


Summary:
Missed one case last time

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23479345](https://our.internmc.facebook.com/intern/diff/D23479345)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support quantize per channel in all cases"


Summary:
Missed one case last time

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23479345](https://our.internmc.facebook.com/intern/diff/D23479345)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support quantize per channel in all cases"


Summary:
Missed one case last time

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23479345](https://our.internmc.facebook.com/intern/diff/D23479345)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support quantize per channel in all cases"


Summary:
Missed one case last time

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23479345](https://our.internmc.facebook.com/intern/diff/D23479345)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support quantize per channel in all cases"


Summary:
Missed one case last time

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23479345](https://our.internmc.facebook.com/intern/diff/D23479345)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support quantize per channel in all cases"


Summary:
Missed one case last time

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23479345](https://our.internmc.facebook.com/intern/diff/D23479345)

[ghstack-poisoned]
